src/Patterns.jl

```diff
@@ -10,101 +10,138 @@ module Patterns
 
 # Examples
 ```jldoctest
-julia> @with_pattern function f end
-f (generic function with 1 method)
+julia> @pattern function f(:one)
+           return 1
+       end
+f (generic function with 2 methods)
+
+julia> @pattern function f(:double, x::Number; plus::Number = 0)::Number
+           return 2x + plus
+       end
+f (generic function with 4 methods)
+```
+
+---
+    @pattern <code>
+
+Resolve pattern.
+
+# Examples
+```jldoctest
+julia> @pattern f(:one)
+1
+
+julia> @pattern f(:double, 123; plus = 1)
+247
+
+julia> @pattern @assert isone(f(:one))
+
+julia> @pattern f(:one) + f(:one) == f(:double, 1)
 ```
 """
-macro with_pattern(f)
-    is_captured = @capture(f, function f_name_ end)
-
-    @assert is_captured
-        "The macro must be called on an empty function declaration!"
-
-    f_name = esc(f_name)
-
-    return quote
-        $(esc(f))
-        function $f_name(args...)
-            _args = symbol_to_val.(args)
-            called_signature =
-                Base.signature_type($f_name, Base.to_tuple_type(typeof(_args)))
-            try
-                if !any(signature -> called_signature <: signature,
-                        patterns($f_name))
-                    throw(MethodError($f_name, _args))
-                end
-
-                return $f_name(_args...)
-            catch e
-                if e isa MethodError && e.f == $f_name
-                    f_name = $f_name
-                    params = typeof(_args).parameters
-                    str = join(params .|> show_pattern_type, ", ")
-                    @error "Missing pattern $f_name($str)"
-                end
-                rethrow(e)
+macro pattern(expr)
+    if is_definition(expr) # Pattern in function definition
+        def::Dict{Symbol, Any} = splitdef(expr)
+
+        # Register pattern
+        register!(def[:name])
+
+        # Replace tokens in signature
+        def[:args] = replace_tokens_in_definition(def[:args])
+
+        expr = combinedef(def)
+
+        return esc(expr)
+    else
+        expr = prewalk(expr) do expr
+            if @capture(expr, f_(args__)) && is_registered(f)
+                args = replace_tokens_in_call(args)
+
+                return :($f($(args...)))
+            else
+                return expr
             end
         end
+
+        return esc(expr)
     end
 end
 
-"""
-    @pattern function definition
+###-----------------------------------------------------------------------------
+### Internals
+###-----------------------------------------------------------------------------
 
-Create a function with symbols as pattern in the signature.
+##------------------------------------------------------------------------------
+## Registered functions
+##------------------------------------------------------------------------------
 
-# Examples
-```jldoctest
-julia> @pattern function f(:one)
-           return 1
-       end
-f (generic function with 2 methods)
+const REGISTERED = Dict{Symbol, Nothing}()
```

Comment:
This doesn't work well across modules.

As a first thought, both `register!` and `is_registered` should be module dependent, i.e., instead of just receiving the name of the function, they should also receive `__module__` from the `@pattern` macro.

In case of `register!`, it can directly use it to construct the qualified name of the function, unless it is an imported function, in which case it should somehow look up the originating module.

In case of `is_registered`, things get even trickier, because it should check the owner module of the symbol `f` at the point of the macro invocation. But that's not possible. I don't know how this could be done properly.

Reply:
I removed this whole mess.